#### Question 1: The above is data for prebid auctions. There are bidders who receive requests for placing ads, they respond to these requests with a bid and the highest bid wins the auction. Analyse the dataset to answer which bidder has the highest win rate (wins / total_bid).

In [1]:
import pandas as pd
import numpy as np


In [82]:
df=pd.read_csv(r"C:\Users\hi\Downloads\analysis.csv")

In [83]:
df.drop("Unnamed: 0",axis=1,inplace=True)

In [84]:
df

,date_hour,client,device_type,time_zone,os_name,browser_name,ad_unit,size,bidder,bid_range,...,sum_time_to_respond,median,min_bid,max_bid,avg_bid,sum_2nd_highest_bid,sum_prebid_winning_bid,sum_winning_bid,sum_nobid,sum_timeout
0,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-Footers,728x90,onemobile,0.5-0.6,...,6614.0,NaN,0.58520,0.58520,0.585200,0.5852,0.5852,NaN,NaN,NaN
1,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-Footers,728x90,rhythmone,0.3-0.4,...,5171.0,NaN,0.33945,0.34875,0.344100,NaN,NaN,NaN,NaN,NaN
2,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-InLine-1,300x250,undertone,0.9-1,...,1523.0,NaN,0.96000,0.96000,0.960000,NaN,NaN,NaN,NaN,NaN
3,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-O-300x250-1,300x250,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-O-300x250-1,300x250,rhythmone,0.3-0.4,...,3570.0,0.34875,0.34410,0.34875,0.346425,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4995,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_20,468x60,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4996,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_20,728x90,medianet,0-0.1,...,363.0,0.05828,0.05828,0.05828,0.058280,NaN,NaN,NaN,NaN,NaN
4997,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_22,468x60,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4998,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_27,310x260,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [85]:
df.columns

Index(['date_hour', 'client', 'device_type', 'time_zone', 'os_name',
       'browser_name', 'ad_unit', 'size', 'bidder', 'bid_range', 'media_type',
       'request_count', 'response_count', 'prebid_win_count', 'win_count',
       'sum_bid', 'sum_time_to_respond', 'median', 'min_bid', 'max_bid',
       'avg_bid', 'sum_2nd_highest_bid', 'sum_prebid_winning_bid',
       'sum_winning_bid', 'sum_nobid', 'sum_timeout'],
      dtype='object')

In [86]:
df["prebid_win_count"]=df["prebid_win_count"].fillna(0)

In [87]:
df

,date_hour,client,device_type,time_zone,os_name,browser_name,ad_unit,size,bidder,bid_range,...,sum_time_to_respond,median,min_bid,max_bid,avg_bid,sum_2nd_highest_bid,sum_prebid_winning_bid,sum_winning_bid,sum_nobid,sum_timeout
0,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-Footers,728x90,onemobile,0.5-0.6,...,6614.0,NaN,0.58520,0.58520,0.585200,0.5852,0.5852,NaN,NaN,NaN
1,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-Footers,728x90,rhythmone,0.3-0.4,...,5171.0,NaN,0.33945,0.34875,0.344100,NaN,NaN,NaN,NaN,NaN
2,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-InLine-1,300x250,undertone,0.9-1,...,1523.0,NaN,0.96000,0.96000,0.960000,NaN,NaN,NaN,NaN,NaN
3,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-O-300x250-1,300x250,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-O-300x250-1,300x250,rhythmone,0.3-0.4,...,3570.0,0.34875,0.34410,0.34875,0.346425,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4995,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_20,468x60,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4996,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_20,728x90,medianet,0-0.1,...,363.0,0.05828,0.05828,0.05828,0.058280,NaN,NaN,NaN,NaN,NaN
4997,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_22,468x60,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4998,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_27,310x260,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [88]:
df["prebid_win_count"].astype('int64')

0       1
1       0
2       0
3       0
4       0
       ..
4995    0
4996    0
4997    0
4998    0
4999    0
Name: prebid_win_count, Length: 5000, dtype: int64

In [90]:
df["win_count"]=df["win_count"].astype('int64')

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

In [91]:
## Analyse the dataset to answer which bidder has the highest win rate (wins / total_bid).
df["win_rate"]=(df["prebid_win_count"]/5000)

In [92]:
df

,date_hour,client,device_type,time_zone,os_name,browser_name,ad_unit,size,bidder,bid_range,...,median,min_bid,max_bid,avg_bid,sum_2nd_highest_bid,sum_prebid_winning_bid,sum_winning_bid,sum_nobid,sum_timeout,win_rate
0,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-Footers,728x90,onemobile,0.5-0.6,...,NaN,0.58520,0.58520,0.585200,0.5852,0.5852,NaN,NaN,NaN,0.0002
1,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-Footers,728x90,rhythmone,0.3-0.4,...,NaN,0.33945,0.34875,0.344100,NaN,NaN,NaN,NaN,NaN,0.0000
2,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-InLine-1,300x250,undertone,0.9-1,...,NaN,0.96000,0.96000,0.960000,NaN,NaN,NaN,NaN,NaN,0.0000
3,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-O-300x250-1,300x250,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0000
4,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-O-300x250-1,300x250,rhythmone,0.3-0.4,...,0.34875,0.34410,0.34875,0.346425,NaN,NaN,NaN,NaN,NaN,0.0000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4995,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_20,468x60,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0000
4996,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_20,728x90,medianet,0-0.1,...,0.05828,0.05828,0.05828,0.058280,NaN,NaN,NaN,NaN,NaN,0.0000
4997,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_22,468x60,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0000
4998,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_27,310x260,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0000


In [93]:
df["win_rate_percent"]=df["win_rate"]*100

In [94]:
win=df[df["win_rate_percent"]!=0]

In [95]:
win["win_rate_percent"].max()

np.float64(2.88)

In [96]:
win[win["win_rate_percent"]==2.0]

,date_hour,client,device_type,time_zone,os_name,browser_name,ad_unit,size,bidder,bid_range,...,min_bid,max_bid,avg_bid,sum_2nd_highest_bid,sum_prebid_winning_bid,sum_winning_bid,sum_nobid,sum_timeout,win_rate,win_rate_percent
1236,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-Responsive-T,728x90,criteo,0-0.1,...,0.011651,0.099264,0.039769,NaN,3.220282,NaN,NaN,NaN,0.02,2.0


In [79]:
win.drop(2528,axis=0,inplace=True)

C:\Users\hi\AppData\Local\Temp\ipykernel_8464\3661055506.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  win.drop(2528,axis=0,inplace=True)


In [101]:
 high=win.groupby(["bidder","win_rate_percent"]).max()

In [103]:
high

KeyError: 0

In [104]:
df

,date_hour,client,device_type,time_zone,os_name,browser_name,ad_unit,size,bidder,bid_range,...,min_bid,max_bid,avg_bid,sum_2nd_highest_bid,sum_prebid_winning_bid,sum_winning_bid,sum_nobid,sum_timeout,win_rate,win_rate_percent
0,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-Footers,728x90,onemobile,0.5-0.6,...,0.58520,0.58520,0.585200,0.5852,0.5852,NaN,NaN,NaN,0.0002,0.02
1,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-Footers,728x90,rhythmone,0.3-0.4,...,0.33945,0.34875,0.344100,NaN,NaN,NaN,NaN,NaN,0.0000,0.00
2,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-InLine-1,300x250,undertone,0.9-1,...,0.96000,0.96000,0.960000,NaN,NaN,NaN,NaN,NaN,0.0000,0.00
3,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-O-300x250-1,300x250,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0000,0.00
4,2021-04-11T01:00:00,asc126828,Desktop,America/Chicago,Linux,Chrome-85.04183133,div-gpt-ad-BT-D-O-300x250-1,300x250,rhythmone,0.3-0.4,...,0.34410,0.34875,0.346425,NaN,NaN,NaN,NaN,NaN,0.0000,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4995,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_20,468x60,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0000,0.00
4996,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_20,728x90,medianet,0-0.1,...,0.05828,0.05828,0.058280,NaN,NaN,NaN,NaN,NaN,0.0000,0.00
4997,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_22,468x60,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0000,0.00
4998,2021-04-11T01:00:00,asc126828,Desktop,America/New_York,Windows,Chrome-89.04389114,div-gpt-ad-BT-D-InPost-8_1_27,310x260,,,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0000,0.00


In [107]:
win_notequal=df[df["prebid_win_count"]!=0]

TypeError: DataFrame.groupby() got multiple values for argument 'axis'

In [111]:
win_rate_df = df.groupby('bidder').agg(
    total_bids=('total_bids', 'sum'),
    wins=('wins', 'sum')
).reset_index()

In [112]:
win_rate_df['win_rate'] = win_rate_df['wins'] / win_rate_df['total_bids']

In [ ]:
win_notequal["prebid_win_count"].sum()

In [113]:
win_rate_df = win_rate_df.sort_values(by='win_rate', ascending=False)

In [ ]:
df['total_bids'] = df['response_count']

In [114]:
win_notequal.groupby(win_notequal["prebid_win_count"].sum(),"bidder").axis

ValueError: No axis named bidder for object type DataFrame

In [ ]:
win_rate_df = win_rate_df.sort_values(by='win_rate', ascending=False)

### undertone has the highest Bid with 0.50 
### Teads has the  2nd highest bid with 0.46

##### Teads has the highesty bid with 2.88%
##### criteo has the 2nd Bid with 2.0%